3p/environment.js

```diff
@@ -0,0 +1,256 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {listenParent} from './messaging';
+
+/**
+ * Info about the current document/iframe.
+ * @type {{inViewport: true}}
+ * @visibleForTesting
+ */
+export const docState = {
+  inViewport: true
+};
+
+let rafId = 0;
+let rafQueue = {};
+
+/**
+ * Add instrumentation to a window and all child iframes.
+ * @param {!Window} win
+ */
+export function manageWin(win) {
+  try {
+    manageWin_(win);
+  } catch (e) {
+    // We use a try block, because the ad integrations often swallow errors.
+    console./*OK*/error(e.message, e.stack);
+  }
+}
+
+/**
+ * @param {!Window} win
+ */
+function manageWin_(win) {
+  if (win.ampSeen) {
+    return;
+  }
+  win.ampSeen = true;
+  // Instrument window.
+  instrumentEntryPoints(win);
+
+  // Watch for new iframes.
+  installObserver(win);
+  // Existing iframes.
+  maybeInstrumentsNodes(win, win.document.querySelectorAll('iframe'));
+}
+
+
+/**
+ * Add instrumentation code to doc.write.
+ * @param {!Window} parent
+ * @param {!Window} win
+ */
+function instrumentDocWrite(parent, win) {
+  const doc = win.document;
+  const write = doc.write;
+  doc.write = function(html) {
+    parent.ampManageWin = function(win) {
+      manageWin(win);
+    };
+    html += '<script>window.parent.ampManageWin(window)</script>';
```

Comment:
I switched to using .close instead. That both makes sure we do it not too often per doc and at a place that more likely has valid HTML.